README.md
# Type-1 Interval Agreement Approach

This repository contains code for applying the type-1 Interval Agreement Approach.

Details of the interval agreement approach are within [From Interval-Valued Data to General Type-2 Fuzzy Sets](https://ieeexplore.ieee.org/document/6762925).

| | From Interval-Valued Data to General Type-2 Fuzzy Sets |
|-|-|
| **Authors** | C. Wagner, S. Miller, J. M. Garibaldi, D. T. Anderson and T. C. Havens |
| **Published in** | IEEE Transactions on Fuzzy Systems, vol. 23, no. 2 April 2015. |
| **Pages** | 248 - 269 |
| **DOI** | [`10.1109/TFUZZ.2014.2310734`](https://doi.org/10.1109/TFUZZ.2014.2310734) |

# Implementations

This repository contains multiple implementations, each in its own directory, with its own specific `README` file.

Here is a brief breakdown of them:

| Implementation | Notes |
| - | - |
| `py` | This is the original Python reference implementation. |
| `js` | This is the ECMAScript Modules package used by the DECSYS Survey Platform. |

# Licensing

## Overview

This software is primarily licensed under the **GNU Affero General Public License v3.0 only** (`AGPL-3.0-only`).

A summary is provided below; the full license text may be found in `LICENSE.md`.

This applies to all implementations in this repository unless otherwise detailed in a specific `README`.

Other license arrangements may be made as appropriate on request.

## Copyright and License Summary

> Type-1 Interval Agreement Approach
>
> Copyright (C) 2015 - 2019 Christian Wagner, LUCID (Lab for Uncertainty in Data and Decision Making)
>
> This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
>
> This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.
>
> You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.


js/CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- ## [Unreleased] -->

## 1.0.2 - 2020-07-25

### Changed

- Security updates from `npm audit fix` and dependabot.

## 1.0.0 - 2019-06-20

### Added

- Azure Pipelines CI/CD config

### Changed

- Various documentation changes

## 1.0.0-beta.5 - 2019-06-10

### Fixed

- fixed the number check for the value `0`. `0` being falsey previously led to this failing.

## 1.0.0-beta.3 - 2019-05-26

### Fixed

- `linspace()` now rounds the number of intervals up to the next largest integer, if it was passed a float

## 1.0.0-beta.2 - 2019-05-24

### Added

- This changelog

### Fixed

- Centroid calculation was hardcoded to between `0` and `10`, with `101` linearly spaced samples.
  - it now uses the smallest and largest values from the dataset
  - and 10 times the width of the dataset for the number of linearly spaced samples.


js/src/iaa.test.js
import IntervalAgreementApproach from "./iaa";

test("matches `main.py`", () => {
  const intervals = [[1, 7], [2, 6], [2, 4], [6, 9]];

  const fs = new IntervalAgreementApproach();
  for (const d of intervals) fs.addInterval(d);

  expect(fs.membership(1)).toEqual(0.25);
  expect(fs.membership(2)).toEqual(0.75);
  expect(fs.membership(7)).toEqual(0.5);
  expect(parseFloat(fs.centroid.toFixed(2))).toEqual(4.57);
});

test("works with floats at both extremes", () => {
  const intervals = [[29, 51.067613404], [20.640571195, 40]];

  const fs = new IntervalAgreementApproach();
  for (const d of intervals) fs.addInterval(d);

  expect(fs.membership(1)).not.toBeNaN();
  expect(parseFloat(fs.centroid)).not.toBeNaN();
});

test("works with values exclusively outside the range 0 - 10", () => {
  const intervals = [[60, 90], [50, 90], [40, 90], [80, 90]];

  const fs = new IntervalAgreementApproach();
  for (const d of intervals) fs.addInterval(d);

  expect(fs.membership(1)).toEqual(0);
  expect(fs.membership(62)).toEqual(0.75);
  expect(fs.membership(42)).toEqual(0.25);
  expect(parseFloat(fs.centroid)).not.toBeNaN();
});

test("works with endpoint only values", () => {
  const intervals = [[0, 100], [0, 0], [100, 100]];

  const fs = new IntervalAgreementApproach();
  for (const d of intervals) fs.addInterval(d);

  expect(fs.membership(1)).not.toBeNaN();
  expect(parseFloat(fs.centroid)).not.toBeNaN();
});


js/src/iaa.js
/**
 * @module
 * This module is for applying the type-1 Interval Agreement Approach.
 *
 * Details of the interval agreement approach are within
 * C. Wagner, S. Miller, J. M. Garibaldi, D. T. Anderson and T. C. Havens,
 * "From Interval-Valued Data to General Type-2 Fuzzy Sets,"
 * in IEEE Transactions on Fuzzy Systems, vol. 23, no. 2,
 * pp. 248-269, April 2015.
 * doi: 10.1109/TFUZZ.2014.2310734
 */

import IntervalMap from "./interval-map";
import linspace from "./linspace";

/**
 * This class type-1 interval agreement approach membership function.
 */
class IntervalAgreementApproach {
  constructor() {
    this.intervals = new IntervalMap();
  }

  /**
   * Add an interval to the fuzzy set.
   * @param {Array<number>} interval in the form `[start, stop]`
   */
  addInterval(interval) {
    this.intervals.set(interval, 1);
  }

  /**
   * Calculate the membership of x.
   * @param {number} x the point to get membership for.
   */
  membership(x) {
    return this.intervals.get(parseFloat(x)) / this.intervals.size;
  }

  /**
   * Return the centroid x-value of the fuzzy set.
   */
  get centroid() {
    const smallestValue = Math.min(...this.intervals.singletonKeys);
    const largestValue = Math.max(...this.intervals.singletonKeys);
    const { top, bottom } = linspace(
      [smallestValue, largestValue],
      (largestValue - smallestValue) * 10 + 1
    ).reduce(
      (a, x) => {
        const mu = this.membership(x);
        return {
          top: a.top + x * mu,
          bottom: a.bottom + mu
        };
      },
      { top: 0, bottom: 0 }
    );
    return top / bottom;
  }
}

export default IntervalAgreementApproach;


js/src/linspace.js
/**
 * @module
 * JS implementation of (most of) Python's `numpy.linspace`
 * since JS doesn't have one in a library.
 *
 * Differences:
 * - interval `start` and `stop` values can only be numbers, not array-like.
 * - `retstep` functionality is not provided.
 * - `dtype` is irrelevant because JS only has the `number` type.
 * - `axis` is irrelevant because we only support `number` not `array` for `start` and `stop`.
 */

/**
 * Returns `num` evenly spaced samples, calculated over the `interval`.
 *
 * The `endpoint` of the interval can optionally be excluded from the results.
 * @param {Array<number>} interval in the form `[start, stop]`.
 * @param {number} [num] Optional number of samples to generate. Deafults to `50`.
 * @param {bool} [endpoint]
 * Whether to include the endpoint (`interval`'s stop value) in the results.
 * Defaults to `true`.
 */
export const linspace = (interval, num = 50, endpoint = true) => {
  if (!(interval instanceof Array) || interval.length !== 2)
    throw new TypeError("Interval must be an Array with 2 elements.");
  if (interval.some(x => isNaN(parseFloat(x))))
    throw new TypeError("Both end points must be numbers.");
  if (typeof num !== "number" || num <= 0)
    throw new TypeError("Num must be a positive integer.");

  //make num an int if it's a float, err on the side of larger
  num = Math.ceil(num);

  // some shortcuts
  const [start, stop] = interval;
  switch (num) {
    case 1:
      return [start];
    case 2:
      return interval;
    default:
      const step = (stop - start) / (endpoint ? num - 1 : num);
      return Array(endpoint ? num : num - 1)
        .fill(start)
        .map((start, i) => start + i * step);
  }
};

export default linspace;


py/interval_dict.py
"""This module is used to create an interval-based dict.

Assign values to interval keys.
For example
d = IntervalDict()
d[1:3] = 2
d[2:4] = 4
# The value of a single-valued key is the maximum assigned value
print(d[1])  # equals 2
print(d[2])  # equals 6
"""

class IntervalDict(object):
    """This class stores a dict in which the keys are intervals."""

    def __init__(self):
        """Initiate the interval dict.

        If a key has been assigned multiple values the sum is returned.
        """
        # store a list of [(interval, value)] mappings
        self._interval_value_pairs = []

    def __setitem__(self, _slice, _value):
        """Assign _value to the continuous _slice.

        _slice must be a slice; e.g. [1:3]
        _value must be numerical
        """
        if not isinstance(_slice, slice):
            raise Exception('The key must be a slice object.')
        if _slice.start is None or _slice.stop is None:
            raise Exception('Both end points must be given')
        self._interval_value_pairs.append(([_slice.start, _slice.stop],
                                           _value))

    def __getitem__(self, _point):
        """Return the value of the singleton _point."""
        values = [0]
        for interval, value in self._interval_value_pairs:
            if self._is_within_interval(interval, _point):
                values.append(value)
        return sum(values)

    def singleton_keys(self):
        """Return the list of key values as singletons."""
        item_list = []
        for interval, value in self._interval_value_pairs:
            if interval[0] not in item_list:
                item_list.append(interval[0])
            if interval[1] not in item_list:
                item_list.append(interval[1])
        return item_list

    def _is_within_interval(self, interval, point):
        """Check if point is within the given interval."""
        if isinstance(interval[0], tuple):
            for s in interval:
                if point >= s[0] and point <= s[1]:
                    return True
        else:
            if point >= interval[0] and point <= interval[1]:
                return True
        return False


js/src/interval-map.js
/**
 * @module
 *
 * This module is used to create an interval-based `Map`.
 *
 * Assign values to interval keys.
 *
 * @example
 * const map = new IntervalMap()
 * map.set([1,3]) = 2
 * map.set([2,4]) = 4
 * // The value of a single-valued key is the maximum assigned value
 * console.log(map.get(1))  // 2
 * console.log(map.get(2))  // 6
 */

/**
 * Store a `Map` in which the keys are intervals
 */
class IntervalMap {
  constructor() {
    /**
     * The Interval Map.
     *
     * If a key has been assigned multiple values the sum is returned.
     *
     * This works in a Map where the keys are reference types (e.g. Array).
     */
    this._map = new Map();
  }

  /**
   * Assign value to the continuous interval key.
   * @param {Array<number>} key
   * The interval as an array in the form `[start, stop]`.
   * @param {number} value The value to assign.
   */
  set(key, value) {
    if (!(key instanceof Array) || key.length !== 2)
      throw new TypeError("The key must be an Array with 2 elements.");
    if (key.some(x => isNaN(parseFloat(x))))
      throw new TypeError("Both end points must be numbers.");
    this._map.set(key, value);
  }

  /**
   * Return the value of the singleton `point`.
   * @param {number} point The point to retrieve the value for.
   */
  get(point) {
    const values = [];
    for (const [interval, value] of this._map) {
      if (this.isWithinInterval(interval, point)) values.push(value);
    }
    return values.reduce((a, v) => a + v, 0);
  }

  /**
   * Return the list of unique interval singleton values
   *
   * @example
   * const map = new IntervalMap();
   * map.set([1,3], 1);
   * map.set([1,5], 1);
   * console.log(map.singletonKeys); // [1, 3, 5]
   */
  get singletonKeys() {
    return [...new Set([...this._map.keys()].reduce((a, v) => [...a, ...v]))];
  }

  get size() {
    return this._map.size;
  }

  /**
   * Check if `point` is within `interval`.
   * @param {Array<number>} interval The interval to check, in the form `[start, stop]`.
   * @param {number} point The point value to check.
   */
  isWithinInterval(interval, point) {
    // py checks if the first element is a tuple
    // I can't work out why it ever would be?
    return point >= interval[0] && point <= interval[1];
  }
}

export default IntervalMap;


py/iaa.py
"""This module is for applying the type-1 Interval Agreement Approach.

Details of the interval agreement approach are within
C. Wagner, S. Miller, J. M. Garibaldi, D. T. Anderson and T. C. Havens,
"From Interval-Valued Data to General Type-2 Fuzzy Sets,"
in IEEE Transactions on Fuzzy Systems, vol. 23, no. 2,
pp. 248-269, April 2015.
doi: 10.1109/TFUZZ.2014.2310734
"""

from numpy import linspace

from interval_dict import IntervalDict


class IntervalAgreementApproach():
    """This class type-1 interval agreement approach membership function."""

    def __init__(self):
        """Create a membership function by the interval agreement approach."""
        self.intervals = IntervalDict()
        self._total_intervals = 0
        self._largest_value = 0  # largest value in the dict after summing
        self.height = 1

    def add_interval(self, interval):
        """Add an interval to the fuzzy set."""
        self.intervals[interval[0]:interval[1]] = 1
        self._total_intervals += 1
        self._largest_value = max([self.intervals[point] for point in
                                   self.intervals.singleton_keys()])
        self.height = self._largest_value / float(self._total_intervals)

    def calculate_membership(self, x):
        """Calculate the membership of x. Returns a float."""
        return float(self.intervals[x]) / self._total_intervals

    def calculate_centroid(self):
        """Calculate the centroid x-value of the fuzzy set."""
        top = 0
        bottom = 0
        for x in linspace(0, 10, 101):
            mu = self.calculate_membership(x)
            top += x * mu
            bottom += mu
        return top / bottom

